<a href="https://colab.research.google.com/github/jorgevillegazz/Estadisticas-y-Machine-Learning-G9---ONE/blob/main/Curso_de_Estad%C3%ADstica_con_Python_1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

***
# <font color=#56B4E9 size=10>CURSO DE ESTADÍSTICA</font>
***

<font color=#56B4E9 size = 5>*Preparando el ambiente*</font>

# <font color=#56B4E9>1 CONOCIENDO LOS DATOS</font>
***

## <font color=#56B4E9>1.1 Dataset del proyecto</font>
***

### Muestra de domicilios Colombia - 2018

Las investigaciones por muestras de domicilios realizadas anualmente, buscan encontrar característcias generales de la población, de educación, trabajo, rendimiento y otras, de acuerdo con las necesidades de información del país, tales como las características de migración, fertilidad, casamientos, salud, nutrición, entre otros temas. Estas muestras al pasar de los años consistuyen una herramienta importante para la formulación, validación y evaluación de políticas dirigidas al desarrollo socioeconómico y la mejora de las condiciones de vida en Colombia.


### Datos

Los datos fueron creados  de manera didáctica para este curso.

### Variables utilizadas

> ### Ingreso
> ***

Ingresos mensuales (en miles de pesos) del trabajo principal para personas de 10 años o más.

> ### Edad
> ***

Edad del entrevistado en la fecha de referencia en años.

> ### Altura
> ***

Altura del entrevistado em metros.

> ### Ciudad
> ***

Código de referência a 27 ciudades analizadas.


> ### Sexo
> ***

|Código|Descripción|
|---|---|
|0|Masculino|
|1|Femenino|

> ### Años de Estudio
> ***

|Código|Descripción|
|---|---|
|1|Sin estudios y menos de 1 año|
|2|1 año|
|3|2 años|
|4|3 años|
|5|4 años|
|6|5 años|
|7|6 años|
|8|7 años|
|9|8 años|
|10|9 años|
|11|10 años|
|12|11 años|
|13|12 años|
|14|13 años|
|15|14 años|
|16|15 años o más|
|17|No se sabe|
||No aplica|

> ### Color
> ***

|Código|Descripción|
|---|---|
|0|Indio|
|2|Blanco|
|4|Negro|
|6|Amarillo|
|8|Moreno|
|9|Sin declarar|

> ### Tratamiento a los datos
***

> Algunos de los tratamientos de datos más frecuentes son:
> 1. Eliminar las observaciones (líneas) con entradas de datos inválidos;
> 2. Eliminar observaciones donde hay datos perdidos (missing data);
> 3. Filtros própios de la investigación, por ejemplo: considerar solo las encuestas realizadas a la cabeza de familia (responsable por el domicilio).

### Importando pandas y leyendo el dataset del projecto

https://pandas.pydata.org/

In [ ]:
import pandas as pd

In [ ]:
datos = pd.read_csv("/content/datos.csv")

In [ ]:
datos.sample(10)

Ciudad  Sexo  Edad  Color  Años de Estudio  Ingreso    Altura
12285      21     0    45      8                9      600  1.746737
41024      33     0    60      8                9     1000  1.581531
45032      33     0    55      2               16    15000  1.803097
72895      52     0    43      2               12     2000  1.574871
41112      33     0    57      2               16     5000  1.642119
51911      35     0    23      2                9     1490  1.697805
15577      23     1    34      8                5     1200  1.643279
55722      41     0    30      2               15      788  1.860847
1058       11     0    27      8               16     2800  1.835925
61049      42     0    30      2               16     4500  1.633864

## <font color=#56B4E9>1.2 Tipos de datos</font>
***

### Clasificación de una variable
<img src='https://caelum-online-public.s3.amazonaws.com/ESP-1752-estadistica-con-python-parte1/img001.png' width='70%'>

### Variables cualitativas ordinales

► Variables que pueden ser ordenadas o que responen algun tipo de jerarquía

In [ ]:
sorted(datos['Años de Estudio'].unique())

[np.int64(1),
 np.int64(2),
 np.int64(3),
 np.int64(4),
 np.int64(5),
 np.int64(6),
 np.int64(7),
 np.int64(8),
 np.int64(9),
 np.int64(10),
 np.int64(11),
 np.int64(12),
 np.int64(13),
 np.int64(14),
 np.int64(15),
 np.int64(16),
 np.int64(17)]

### Variables cualitativas nominales

► Variables que no pueden ser ordenadas ni responden alguna jerarquía.

In [ ]:
sorted(datos['Sexo'].unique())

[np.int64(0), np.int64(1)]

In [ ]:
sorted(datos['Color'].unique())

[np.int64(0), np.int64(2), np.int64(4), np.int64(6), np.int64(8)]

In [ ]:
sorted(datos['Ciudad'].unique())

[np.int64(11),
 np.int64(12),
 np.int64(13),
 np.int64(14),
 np.int64(15),
 np.int64(16),
 np.int64(17),
 np.int64(21),
 np.int64(22),
 np.int64(23),
 np.int64(24),
 np.int64(25),
 np.int64(26),
 np.int64(27),
 np.int64(28),
 np.int64(29),
 np.int64(31),
 np.int64(32),
 np.int64(33),
 np.int64(35),
 np.int64(41),
 np.int64(42),
 np.int64(43),
 np.int64(50),
 np.int64(51),
 np.int64(52),
 np.int64(53)]

### Variables cuantitativas discretas

► Variables que representan un conteo donde los valores posibles forman un conjunto finito o numerable.


In [ ]:
datos.Edad.max()

99

In [ ]:
print(f"La edad mínima es {datos.Edad.min()} y la edad máxima es {datos.Edad.max()}")

La edad mínima es 13 y la edad máxima es 99


#### <font color='red'>Observación</font>
***
> La variable edad puede ser clasificada de tres formas diferentes:
> 1. <b>CUANTITATIVA DISCRETA</b> - cuando representa años completos (números enteros);
> 2. <b>CUANTITATIVA CONTINUA</b> - cuando representa la edad exacta, siendo representada por fracciones de años;
> 3. <b>CUALITATIVA ORDINAL</b> - cuando representa  intervalos de edad.

### Variables cuantitativas continuas

► Variables que representan un conteo o medición que asume valores en escalas continuas (números reales).

In [ ]:
print(f"La Altura mínima es {datos['Altura'].min()} y la altura máxima es {datos['Altura'].max()}")

La Altura mínima es 1.339244614 y la altura máxima es 2.028496765


# <font color=#56B4E9>2 DISTRIBUCIÓN DE FRECUENCIAS</font>
***

El primer paso en un análisis es conocer el comportamiento de las variables involucradas en el estudio. Utilizando técnicas estadísticas como el análisis de las <b> DISTRIBUCIONES DE FRECUENCIAS </b> e <b> HISTOGRAMAS </b> podemos evaluar mejor cómo se distribuyen los fenómenos del estudio.



## <font color=#56B4E9>2.1 Distribución de frecuencias para variables cualitativas</font>
***

### Método 1

https://pandas.pydata.org/pandas-docs/version/0.22/generated/pandas.Series.value_counts.html

In [ ]:
#Para acceder  a nuestros Datos de la colimna Sexo  y  ver LA FRECUENCIA cuantos tenemos de masculiono y femenino
datos.Sexo.value_counts()

Sexo
0    53250
1    23590
Name: count, dtype: int64

In [ ]:
# Ahor a queremos  ver  el Porcentaje  de estos Datos  (Sexo)
datos.Sexo.value_counts(normalize = True).round(4) * 100   # se multiplica  por 100  para  que nos de  lso vales  en tanto porciento
#                                                             y la funcion round para  que despliege  las decimales que queramos

Sexo
0    69.3
1    30.7
Name: proportion, dtype: float64

In [ ]:
# Depositamos este resultado en una variable
frecuencia_sexo=datos.Sexo.value_counts()
frecuencia_sexo

Sexo
0    53250
1    23590
Name: count, dtype: int64

In [74]:
#ahora  usaremos otra variable llamada  procentaje
porcentaje_sexo=datos.Sexo.value_counts(normalize = True).round(4) * 100
porcentaje_sexo

Sexo
0    69.3
1    30.7
Name: proportion, dtype: float64

Y ahora queremos ver nuestra distribución de frecuencia, Nuestra tabla. ¿Cómo hacemos para ver nuestra tabla?

Para crear nuestra tabla, necesitamos crear un nuevo DataFrame. Y para crear un nuevo DataFrame, necesitamos llamar a nuestro amigo Pandas. Para eso existe una función que decimos así: Pandas, (pd) Crea para mí, por favor, un DataFrame.

In [75]:
pd.DataFrame({'Frecuencia':frecuencia_sexo,'Porcentaje %':porcentaje_sexo})

Frecuencia  Porcentaje %
Sexo                          
0          53250          69.3
1          23590          30.7

En elcodigo  anterior dentro del parentesis se crean dos COLUMNAS de nuetra nueva tabla " Frecuencia y Porcentaje % " las  cuales su contenido  son los valores de las variables frecuencia_sexo y porcentaje_sexo respectivamente.

 ¿Quedó bien? Pero, y si llamamos de nuevo nuestra tabla, no va a ver cómo llamarla. Entonces, aquí vamos a atribuir esto aquí, en una nueva variable, que va a ser dist_frec_cualitativa

In [76]:
dist_frec_cualitativa=pd.DataFrame({'Frecuencia':frecuencia_sexo,'Porcentaje %':porcentaje_sexo})
dist_frec_cualitativa

Frecuencia  Porcentaje %
Sexo                          
0          53250          69.3
1          23590          30.7

Oh, no sé si se han dado cuenta que Sexo es 0 y 1. Pero si ustedes van a,  mostrar esto para una persona que es líder de ustedes, él no va a entender que es 0 y 1. Entonces, vamos a tener que **renombrar** nuestros valores aquí. ¿Y cómo nosotros renombramos? Primero, nosotros tenemos que pensar que esta parte de aquí es un **Axis**, ¿no? Entonces, aquí está dentro del Axis, que va a ser de columna.

In [78]:
dist_frec_cualitativa.rename_axis('Sexo',axis='columns', inplace=True) #cambiamos el axis a columna sexo
dist_frec_cualitativa.rename(index={1:'Femenino',0:'Masculino'},inplace=True) #cambiamos el nombre de los indices

In [79]:
dist_frec_cualitativa

Sexo       Frecuencia  Porcentaje %
Sexo                               
Masculino       53250          69.3
Femenino        23590          30.7

### Método 2

https://pandas.pydata.org/pandas-docs/version/0.22/generated/pandas.crosstab.html

Ya está correctamente nuestra tabla.

Sin embargo, tenemos **otro método** que nos permite ver mejor.

¿Cómo lo conseguimos?Aquí de nuevo tenemos una documentación que podemos usar, que es utilizando el **crosstab**.¿Y qué significa el crosstab?

**La función crosstab en pandas** sirve para **calcular tablas de frecuencia cruzada (contingencia) de dos o más factores**, lo que permite analizar relaciones, patrones y tendencias entre variables categóricas. Genera un DataFrame que resume cómo se distribuyen los datos, mostrando frecuencias absolutas o porcentajes.

**Principales usos y características:**

*Análisis de frecuencia*: **Por defecto**, cuenta cuántas veces ocurren las combinaciones de categorías (frecuencia).

*Tabla de Contingencia*: Ideal para cruzar datos como encuestas o estudios de mercado.

*Normalización*: Con normalize=True, **calcula porcentajes** en lugar de conteos absolutos.

*Márgenes:* Permite **añadir filas/columnas de totales** usando margins=True.

*Agregación:* Puede **calcular estadísticas (media, suma, etc.)** sobre una tercera variable numérica usando values y aggfunc.


Nosotros podemos, de alguna manera, ver dos variables al mismo tiempo .Por ejemplo, estamos viendo color y sexo, o podríamos ver otra variable que también sea importante.solo que creé un diccionario que clasifica masculino y femenino para sexo, y para color con categorías como indígena, blanco, negro, amarillo, etc.Entonces, ¿cómo empezamos a usar el método del crosstab?

In [ ]:
# Creamos  un par de  Dicionarios, solo para llegar a cambiar el nombre   mas adelante de la  tabla
sexo = {0: 'Masculino',
        1: 'Femenino'}

color = {0: 'Indigena',
         2: 'Blanco',
         4: 'Negro',
         6: 'Amarillo',
         8: 'Pardo',
         9: 'Sin declarar'}

In [ ]:
#Creamo una  nueva variable(tabla) y ...
# Extraemos los datos de la tabla(datos) en este caso los de las  columnas  Sexo y Color
frecuencia_cross =pd.crosstab(datos.Sexo, datos.Color)
frecuencia_cross

Color    0      2     4    6      8
Sexo                               
0      256  22194  5502  235  25063
1      101   9621  2889  117  10862

Una vez  ejecutado  apreciamos  la nueba tabla con los datos " CRUZADOS"

Solo que tanto los  nombres de indices y columnas un usiario no los podría interpretar.

Es por eso  que  les cambiaremos  los nombres  para facilitar su entendimiento.

In [ ]:
# Ahora  si  usaremos  los dos diccionarios que creamos anteriormente (sexo y color)
frecuencia_cross.rename(index=sexo,inplace=True)
frecuencia_cross.rename(columns=color,inplace=True)
frecuencia_cross

Color      Indigena  Blanco  Negro  Amarillo  Pardo
Sexo                                               
Masculino       256   22194   5502       235  25063
Femenino        101    9621   2889       117  10862

Supongamos que  quiero ver también cómo se ve la proporción, además de los valores absolutos, cómo hacer?.

In [ ]:
porcentaje2 = pd.crosstab(datos.Sexo,
                         datos.Color,
                         normalize=True).round(4)*100

porcentaje2.rename(index = sexo,inplace = True)
porcentaje2.rename(columns = color,inplace = True)
porcentaje2

Color      Indigena  Blanco  Negro  Amarillo  Pardo
Sexo                                               
Masculino      0.33   28.88   7.16      0.31  32.62
Femenino       0.13   12.52   3.76      0.15  14.14

En el mundo del análisis de datos, a menudo es necesario explorar la relación entre dos variables categóricas. El código que hemos visto utiliza la función pd.crosstab de Pandas para crear una tabla de contingencia(cruzada), que nos **permite examinar cómo se distribuyen los valores de una variable** (*Ingreso en este caso*) en función de las combinaciones de otras dos variables categóricas (Sexo y Color).

In [ ]:
porcentaje_ingreso = pd.crosstab(datos.Sexo,
                         datos.Color,
                         aggfunc = 'mean',
                         values = datos.Ingreso)

porcentaje_ingreso.rename(index = sexo,inplace = True)
porcentaje_ingreso.rename(columns = color,inplace = True)
porcentaje_ingreso

Color         Indigena       Blanco        Negro     Amarillo        Pardo
Sexo                                                                      
Masculino  1081.710938  2925.744435  1603.861687  4758.251064  1659.577425
Femenino   2464.386139  2109.866750  1134.596400  3027.341880  1176.758516

# **Explicación paso a paso:**

* pd.crosstab(datos.Sexo, datos.Color...): Esta es la función principal que crea la tabla de contingencia.

* Los dos primeros argumentos (datos.Sexo y datos.Color) especifican las dos variables categóricas que se cruzarán.

* aggfunc = 'mean': Indica que queremos calcular la media de datos.Ingreso para cada combinación de Sexo y Color. Podríamos usar otras funciones de agregación como sum, count, etc., dependiendo del análisis que queramos realizar.

* values = datos.Ingreso: Especifica la variable numérica cuyos valores se agregarán en cada celda de la tabla.

* porcentaje.rename(index = sexo, inplace = True): Esta línea renombra las etiquetas del índice (filas) de la tabla de contingencia utilizando los valores del diccionario sexo. El argumento inplace = True modifica la tabla directamente en lugar de crear una nueva.

* porcentaje.rename(columns = color, inplace = True): De manera similar, esta línea renombra las etiquetas de las columnas de la tabla utilizando los valores del diccionario color.

* porcentaje: Finalmente, se muestra la tabla resultante.

# ¿Qué nos muestra esta tabla?

La tabla resultante nos mostrará ***la media de Ingreso para cada combinación de Sexo y Color***. Esto nos permite comparar, por ejemplo, si hay diferencias significativas en los ingresos promedio entre hombres y mujeres de diferentes etnias.

Hay que experimentar con diferentes funciones de agregación y variables en pd.crosstab **para descubrir patrones interesantes en tus datos**.

¡**El análisis bivariado** es una herramienta poderosa para comprender mejor las relaciones entre variables en tus conjuntos de datos! ;))

## <font color=#56B4E9>2.2 Distribución de frecuencia para variables cuantitativas (clases personalizadas)</font>
***

Vamos a ver nuestras variables y crear nuestras tablas de frecuencias para variables cuantitativas, ya que hemos visto las cualitativas.

Entonces, el primero paso que nosotros vamos a hacer, dado que vamos a usar nuestra variable cuantitativa, que ahora va a ser nuestra variable **Ingreso**, nosotros, como primer paso, es **especificar los límites de cada clase**.

Entonces, nosotros tenemos que el salario mínimo, por ejemplo, en el momento que nosotros estamos viendo la base de datos en 2018, como un dato bien así, que hemos usado ahí dentro de n**uestra tabla del censo del 2018**, que fue usado en Colombia, fue la última vez que se ha hecho un censo, **el salario mínimo ahí era de 788.000** pesos colombianos.

Dicho esto, si 788.000 pesos colombianos eran en 2018, lo que nosotros tenemos que ver,  para **categorizar nuestras clases** tenemos que A es más de 20 salarios mínimos, B es de 10 a 20 salarios mínimos, C de 4 a 10, 2 a 4 y así sucesivamente.

Entonces, si es 788.000, es 788.000 veces 20, 788.000 veces 4, 10 y por ahí va, hasta que nosotros tengamos aquí nuestros valores que nosotros queremos.Entonces, tenemos que la categoría A va a ser de más de 15.760, B de 7.880.000 a 15.760 y así sucesivamente con ese cálculo,

### Paso 1 - Especifique los límites de cada clase

Utilice la siguiente clasificación:

<b>A</b> ► Más de 20 SM

<b>B</b> ► De 10 a 20 SM

<b>C</b> ► De 4 a 10 SM

<b>D</b> ► De 2 a 4 SM

<b>E</b> ► Hasta 2 SM

donde <b>SM</b> es el valor del salario mínimo en ese momento. En nuestro caso <b>788 mil pesos colombianos</b> (2018):

<b>A</b> ► Más de 15.760

<b>B</b> ► De 7.880 a 15.760

<b>C</b> ► De 3.152 a 7.880

<b>D</b> ► De 1.576 a 3.152

<b>E</b> ► Hasta 1.576


Entonces, ¿qué es lo que vamos a hacer ahora? Primero, vamos a explorar un poco más nuestra variable de ingreso, entonces va a ser datos.Ingreso.min(), ¿cuál es el salario mínimo? Cero, está bien.


In [56]:
datos.Ingreso.min()

0


Y ahí nosotros queremos saber cuál es el salario máximo de la persona que ha sido entrevistada. Va a ser datos, de nuevo, entrar a nuestro dataset, datos.Ingreso.max(), el valor máximo. Y tenemos que el valor máximo aquí es 200.000 pesos colombianos. ¡Belleza! Tenemos eso.

In [57]:
datos.Ingreso.max()

200000

Entonces, **clases** = va a ser igual de 0 a 1.576, de 1.576 a 3.152, después de 3.152 a 7.880, y así sucesivamente hasta llegar a nuestro valor máximo que es 200.000.

Ahora vamos a colocar nuestros <font color= blue>**labels**</font>. ¿Qué son nuestros labels? Nada más que los nombres de las clases que él va a acabar haciendo.

 Entonces, imaginen que cada una de estas clases es un intervalo, Entonces, para este intervalo, el label será el nombre A,B...,Esto es lo que significa label.


In [58]:
clases=[0,1576,3152,7880,15760,200000]

Colocaremos aquí nuestra variable label, que será la siguiente. Va a ser ['E', 'D', 'C', 'B', 'A'] , de acuerdo al orden que hemos establecido

In [61]:
label = ['E', 'D', 'C', 'B', 'A']

### Paso 2 - Crear la tabla de frecuencias

¿Qué haremos en nuestro segundo paso? <font color=red>Crear nuestra tabla de frecuencia</font>

Para usar una tabla de frecuencia para datos cuantitativos, vamos a usar una función muy importante: <font color=blue>pd.cut</font>.


La función **pandas.cut()** se utiliza para convertir variables numéricas continuas en variables categóricas o discretas, agrupando los datos **en contenedores ("bins") o intervalos específicos**.

Es fundamental para transformar datos numéricos sesgados, segmentar rangos de edades, precios, o facilitar análisis de datos y modelos de machine learning. **

In [64]:
frec_var_cuant = pd.cut(x = datos['Ingreso'],
                         bins=clases,
                         labels=label,
                        include_lowest=True)

Creamo una  variable  para  nuestra tabla de frecuencias.

Primero, pd.cut, y estamos diciendo que x será nuestra base de datos. Esa base de datos será simplemente datos.

¿Cuál es la variable que estamos viendo? En este caso será Ingreso

También queremos añadir un parámetro que serán nuestros **bins**, que son las clases, los intervalos que hemos creado.

**Label**, que aquí es muy redundante, Un parámetro muy importante

 aquí es que queremos incluir el valor de cero, y para eso usamos **include_lowest**. Por defecto, viene como falso, pero aquí lo pondremos como True.

In [63]:
frec_var_cuant

0        E
1        E
2        E
3        C
4        E
        ..
76835    E
76836    E
76837    E
76838    E
76839    E
Name: Ingreso, Length: 76840, dtype: category
Categories (5, object): ['E' < 'D' < 'C' < 'B' < 'A']

Entonces, ¿cuántos E tenemos? ¿Cuántos C tenemos? ¿Cuántos A tenemos? Para hacer eso, ahora necesitamos usar **pandas.value_counts**.



In [72]:
frec_var_cuant = pd.value_counts(
                                pd.cut(x = datos['Ingreso'],
                                 bins=clases,
                                 labels=label,
                                include_lowest=True))

/tmp/ipython-input-4055139441.py:1: FutureWarning: pandas.value_counts is deprecated and will be removed in a future version. Use pd.Series(obj).value_counts() instead.
  frec_var_cuant = pd.value_counts(


In [73]:
frec_var_cuant

Ingreso
E    49755
D    16700
C     7599
B     2178
A      608
Name: count, dtype: int64

Ahora sí, aquí se da frecuencia, oh, voilà. Entonces, tenemos que en E hay 49.755 personas, En la categoría D, son 16.700, y así sucesivamente,  Entonces, esto aquí lo vamos a dejar solo para organizar, en este código, para que esté mejor.

Ahora que tenemos nuestros valores absolutos, ¿qué es lo que queremos ver también? ¿Por qué no ver porcentajes?

Es lo mismo codigo  solo que  agregamos  normalize, solo que tiene que estar dentro de los paréntesis de value_counts, no dentro de los de pd.cut

In [83]:
porcentaje_var_cuant = pd.value_counts(
                                pd.cut(x = datos['Ingreso'],
                                 bins=clases,
                                 labels=label,
                                include_lowest=True),
                                normalize=True).round(4)*100


/tmp/ipython-input-4006915082.py:1: FutureWarning: pandas.value_counts is deprecated and will be removed in a future version. Use pd.Series(obj).value_counts() instead.
  porcentaje_var_cuant = pd.value_counts(


In [84]:
porcentaje_var_cuant

Ingreso
E    64.75
D    21.73
C     9.89
B     2.83
A     0.79
Name: proportion, dtype: float64

Muy bien , ahora  juntaremos  nustras  dos  tablas  para  hacer una sola como se  hizo en las frecuancias de variables cualitativsa

In [87]:
dist_frec_cuantitativa_personalizado=pd.DataFrame({'Frecuencia':frec_var_cuant,'Porcentaje %':porcentaje_var_cuant})

In [88]:
dist_frec_cuantitativa_personalizado

Frecuencia  Porcentaje %
Ingreso                          
E             49755         64.75
D             16700         21.73
C              7599          9.89
B              2178          2.83
A               608          0.79

Solo que queremos ver, por ejemplo, de una ***manera organizada***. ¿Cómo conseguimos ver eso? Existe una función que sería sort_index, y aquí sería ascending, o sea, de manera ascendente, y va a ser false.

In [89]:
dist_frec_cuantitativa_personalizado.sort_index(ascending=False)

Frecuencia  Porcentaje %
Ingreso                          
A               608          0.79
B              2178          2.83
C              7599          9.89
D             16700         21.73
E             49755         64.75

## <font color=#56B4E9>2.3 Distribución de frecuencia para variables cuantitativas (clases de amplitud fija)</font>
***

Ahora, tenemos esta distribución de frecuencias personalizadas cuantitativas. Si queremos algo que tenga una amplitud fija, existe una teoría que se llama la <font color = green>**regla de Sturges**</font>.

Para eso, ¿qué es lo que sugiere? Imaginen que están organizando su biblioteca en casa, solo que tienen una inmensa cantidad de libros, y necesitan saber cuántos estantes van a querer para su biblioteca. Para saber este número de estantes, hacen la regla de Sturges. Él va a definir correctamente un estante donde tus libros no queden ni muy apretados ni estén danzando ahí en su estante. Básicamente, lo que hace es este cálculo

***

La ***regla de Sturges*** es un criterio estadístico utilizado para determinar el número óptimo de intervalos o clases (**k**) necesarios para organizar un conjunto de datos en un histograma o tabla de frecuencias. Propuesta por Herbert Sturges en 1926, equilibra el detalle de los datos con su interpretabilidad

### Importando bibliotecas

http://www.numpy.org/

### Paso 1 - Difiniendo el número de clases

#### Regla de Sturges

# $$k = 1 + \frac {10}{3}\log_{10}n$$

### Paso 2 - Crear la tabla de frecuencias

## <font color=#56B4E9>2.4 Histograma</font>
***

El <b>HISTOGRAMA</b> es la representación gráfica de una distribución de frecuencia. Es un gráfico formado por un conjunto de rectángulos colocados uno al lado del otro, donde el área de cada rectángulo es proporcional a la frecuencia de la clase que representa.

### Importando la biblioteca

https://seaborn.pydata.org/

# <font color=#56B4E9>3 MEDIDAS DE TENDENCIA CENTRAL</font>
***

## DataFrame de ejemplo

In [ ]:
df = pd.DataFrame(data = {'María': [8, 10, 4, 8, 6, 10, 8],
                          'Pedro': [10, 2, 0.5, 1, 3, 9.5, 10],
                          'Pablo': [7.5, 8, 7, 8, 8, 8.5, 7]},
                  index = ['Matemática',
                           'Portugués',
                           'Inglés',
                           'Geografía',
                           'Historia',
                           'Física',
                           'Química'])
df.rename_axis('Asignaturas', axis = 'columns', inplace = True)
df

Asignaturas  María  Pedro  Pablo
Matemática       8   10.0    7.5
Portugués       10    2.0    8.0
Inglés           4    0.5    7.0
Geografía        8    1.0    8.0
Historia         6    3.0    8.0
Física          10    9.5    8.5
Química          8   10.0    7.0

## <font color=#56B4E9>3.1 Media aritmética</font>
***

Es representada por $\mu$ cuando se refiere a la población y por $\bar{X}$ cuando se refiere a la muestra

# $$\mu = \frac 1n\sum_{i=1}^{n}X_i$$

donde

$n$ = número de observaciones (registros)

$X_i$ = valor de la i-ésima observación (registro)

## <font color=#56B4E9>3.2 Mediana</font>
***

Para obtener la mediana de un conjunto de datos, debemos proceder de la siguiente manera:
1. Ordenar el conjunto de datos;
2. Identificar el número de observaciones (registros) del conjunto de datos ($n$);
3. Identicar el elemento del medio:

> Cuando $n$ sea impar, la posición del elemento del medio se obtendrá de la siguiente manera:


# $$Elemento_{Md} = \frac{n+1}{2}$$

> Cuando $n$ sea par, la posición del elemento mediano se obtendrá de la siguiente manera:


# $$Elemento_{Md} = \frac{n}2$$

4. Obtener la mediana:

> Cuando $n$ sea impar:


# $$Md = X_{Elemento_{Md}}$$

> Cuando $n$ sea par:


# $$Md = \frac{X_{Elemento_{Md}} + X_{Elemento_{Md}+1}}2$$
***

### Ejemplo 1 - n impar

<img src='https://caelum-online-public.s3.amazonaws.com/ESP-1752-estadistica-con-python-parte1/img002.png' width='40%' style="float:left">

### Ejemplo 2 - n par

<img src='https://caelum-online-public.s3.amazonaws.com/ESP-1752-estadistica-con-python-parte1/img003.png' width='50%' style="float:left">

### Obtener la mediana en nuestro conjunto de datos

## <font color=#56B4E9>3.3 Moda</font>
***

La moda se puede definir como el valor más frecuente de un conjunto de datos. La moda es ampliamente utilizada para datos cualitativos.

### Obteniendo la moda de nuestro dataset

## <font color=#56B4E9>3.4 Relación entre media, mediana e moda</font>
***

<img src='https://caelum-online-public.s3.amazonaws.com/ESP-1752-estadistica-con-python-parte1/img004.png' width='80%'>

### Evaluando la variable INGRESO

***

### Evaluando la variable ALTURA

***

### Evaluando la variable AÑOS DE ESTUDIO

# <font color=#56B4E9>4 MEDIDAS DE LOCALIZACIÓN</font>
***

## <font color=#56B4E9>4.1 Cuartiles, deciles y percentiles</font>
***

Hay una serie de medidas de posición similares en su diseño a la mediana, aunque no son medidas de tendencia central. Como se sabe, la mediana divide la distribución en dos partes iguales en términos del número de elementos en cada parte. Los cuartiles, por otro lado, permiten que la distribución se divida en cuatro partes iguales en términos del número de elementos en cada uno; deciles en diez partes y centiles en cien partes iguales.

## <font color=#56B4E9>4.2 Box-plot</font>
***

El *box plot* da una idea de la posición, dispersión, asimetría, colas y valores extremos. La posición central está dada por la mediana y la dispersión por $IIQ$ (distancia inter cuartil). Las posiciones relativas de $Q1$, $Mediana$ y $Q3$ dan una idea de la simetría de la distribución. Las longitudes de las colas están dadas por las líneas que van desde el rectángulo a los valores remotos y por los valores extremos.

<img src='https://caelum-online-public.s3.amazonaws.com/ESP-1752-estadistica-con-python-parte1/img005.png' width='65%'>

<img src='https://caelum-online-public.s3.amazonaws.com/ESP-1752-estadistica-con-python-parte1/img006.png' width='80%'>

# <font color=#56B4E9>5 MEDIDAS DE DISPERSIÓN</font>
***

Aunque las medidas de tendencia central y de localización proporcionan un resumen muy importante de los datos, pueden no ser suficientes para caracterizar diferentes conjuntos, especialmente cuando las observaciones de una distribución dada presentan datos muy dispersos.

## <font color=#56B4E9>5.1 Desviación media absoluta</font>
***


# $$DM = \frac 1n\sum_{i=1}^{n}|X_i-\bar{X}|$$


## <font color=#56B4E9>5.2 Varianza</font>
***

### Varianza

La varianza se construye a partir de las diferencias entre cada observación y la media de los datos, es decir, la desviación alrededor de la media. Al calcular la varianza, las desviaciones alrededor de la media son elevadas al cuadrado.

### Varianza de la población

# $$\sigma^2 = \frac 1n\sum_{i=1}^{n}(X_i-\mu)^2$$

### Varianza de la muestra

# $$S^2 = \frac 1{n-1}\sum_{i=1}^{n}(X_i-\bar{X})^2$$

## <font color=#56B4E9>5.3 Desviación estándar</font>
***

Una de las restricciones de la varianza es el hecho de que proporciona medidas cuadráticas de las unidades originales; la varianza de las medidas de longitud, por ejemplo, está en unidades de área. Por lo tanto, el hecho de que las unidades sean diferentes dificulta la comparación de la dispersión con las variables que la definen. Una forma de eliminar esta dificultad es considerar su raíz cuadrada.

### Desviación estándar de la población

# $$\sigma = \sqrt{\frac 1n\sum_{i=1}^{n}(X_i-\mu)^2} \Longrightarrow \sigma = \sqrt{\sigma^2}$$


### Desviación estándar de la muestra

# $$S = \sqrt{\frac 1{n-1}\sum_{i=1}^{n}(X_i-\bar{X})^2} \Longrightarrow S = \sqrt{S^2}$$